Migrate to ocamlmerlin-lsp

rusty-key opened the pull request:
Original PR is here:
https://github.com/reasonml-editor/vscode-reasonml/pull/264

Native LSP implementation was merged recently into merlin repository. This PR switches extension to use it instead of OLS. This goes with some features removed (temporarily, I hope):

Also, it adds support for `ocamlformat`/`refmt`, because it was handled by OLS.

What do you think? What else should be done here?

andreypopp commented:
Could you write a summary of what's removed and what's added here?

andreypopp commented on src/formatters/ocaml.ts:
```diff
@@ -0,0 +1,37 @@
+import { execSync } from "child_process";
```
Same comments I left for reason formatter are applicable to this too, I think.

andreypopp commented on src/client/index.ts:
```diff
@@ -29,29 +36,103 @@ class ErrorHandler {
+  return serverOptions;
+}
+
+export async function launchMerlinLsp(context: vscode.ExtensionContext, useEsy: boolean): Promise<void> {
```
Let change `useEsy` arg to be `options: {useEsy: bool}` instead, this will make things more clear on call sites.

rusty-key commented on src/formatters/reason.ts:
```diff
@@ -0,0 +1,33 @@
+          fs.writeFileSync(tempFileName, textEditor.document.getText(), "utf8");
+          const formattedText = execSync(`${formatter} ${tempFileName}`).toString();
+          const textRange = getFullTextRange(textEditor);
+          fs.unlinkSync(tempFileName);
```
I tried it, but it doesn't work with large files

rusty-key commented:
@andreypopp, I edited the description to mention removed features and addressed your comments

I guess I also need to provide a guide on how to build `ocamlmerlin-lsp`?

bryphe commented on src/formatters/ocaml.ts:
```diff
@@ -0,0 +1,37 @@
+        const textEditor = vscode.window.activeTextEditor;
+
+        if (textEditor) {
+          const tempFileName = `/tmp/vscode-reasonml-${uuidv4()}.ml`;
```
This should be using `os.tmpdir()` instead of `/tmp` - otherwise it won't work correctly on Windows.

rusty-key replied:
I didn't think about windows at all. Will try to setup vm and test it there this week

bryphe replied:
Thanks @rusty-key !

andreypopp commented:
from this list:
> — find occurances;
— symbol rename;
— case splitting;
— code lens.

only "case splitting" is still unsupported by merlin-lsp, others were implemented in ocamlmerlin-lsp

jchavarri commented on package.json:
```diff
@@ -16,138 +16,57 @@
-        "reason.diagnostics.tools": {
-          "type": "array",
-          "items": {
-            "enum": ["merlin", "bsb"]
```
@rusty-key @andreypopp One of the main advantages of this bsb integration was that bsb has full project visibility. So if one file changes and that change breaks other files, bsb is able to show those errors in the Diagnostics panel, even if the file is not opened. One has just to click on them to open the broken file, which is useful for large refactors.

Is something like that possible with merlin? I always believed that merlin only has a "local view" of the world. 🤔

jchavarri commented:
Another question: what's the best way to keep compatibility with `ocamlmerlin` and BuckleScript projects? `ocamlmerlin` is a crucial part of this extension to jump to definition etc, but the historical way to do so (using `reason-cli`) hasn't been updated since last August.

Ideally, this extension could be used to run both BuckleScript and native projects, but I wonder what's the best way to use merlin with BS projects, considering it doesn't play well with esy yet? cc @andreypopp

andreypopp commented:
@jchavarri I've been managing native dependencies with esy in bs projects so far — `esy.json` for esy and `package.json` for yarn/bs. It worked fine.

rusty-key commented:
I included precompiled (with `ocaml-4.02.3+buckle`) versions of `merlin-lsp` and `merlin-reason`. Now extension should work with Bucklescript projects on mac and linux out-of-the-box. 

Next steps are:
— do the same for windows (if it is possible)
— provide setup instructions for non-bs projects
— automate setup for non-bs projects.

jchavarri commented on src/client/index.ts:
```diff
@@ -28,30 +35,157 @@ class ErrorHandler {
+  if (options.useEsy) {
+    run = {
+      args: ["exec-command", "--include-current-env", merlinLsp],
+      command: process.platform === "win32" ? "esy.cmd" : "esy",
```
The way this is implemented has some unexpected behavior. If I'm understanding correctly, if user sets in `settings.json`:

```json
{
  "reason.path.ocamlmerlin-lsp": "/some/folder/ocamlmerlin-lsp"
}
```

then the command will be
```bash
esy /some/folder/ocamlmerlin-lsp
```

but I would expect it to be just 

```bash
/some/folder/ocamlmerlin-lsp
```

If the path is set manually, then it should be used "as is", imo.

jchavarri commented on package.json:
```diff
@@ -16,138 +16,57 @@
-          "default": "./node_modules/bs-platform/lib/bsb.exe",
-          "description": "The path to the `bsb` binary."
-        },
-        "reason.path.ocamlfind": {
```
@jordwalke Heads up: this will remove the APIs for introspecting environments that were added in https://github.com/freebroccolo/ocaml-language-server/pull/68.

jchavarri commented on package.json:
```diff
@@ -16,138 +16,57 @@
-          "default": "ocamlfind",
-          "description": "The path to the `ocamlfind` binary."
-        },
         "reason.path.esy": {
```
@rusty-key I think this can be removed too, it was used for the env introspection (like `ocamlfind`, `env` etc)

jchavarri commented on src/client/index.ts:
```diff
@@ -28,30 +35,157 @@ class ErrorHandler {
+        env: {
+          ...process.env,
+          MERLIN_LOG: "-",
+          OCAMLFIND_CONF: "/dev/null",
```
Does this have any implications? Why does merlin fail if this is not set? cc @andreypopp

rusty-key replied:
afaiu, merlin relies on `ocamlfind` by default and when it is not there, it fails. This flag can be removed as soon as `findless` branch will be merged and released:
https://github.com/ocaml/merlin/tree/findless

jchavarri commented:
One thing I'm noticing when testing this branch is that when opening `.re` files with comments like `/* */`, merlin seems to choke:

```
# 0.02 lsp - debug
send: {
  "jsonrpc": "2.0",
  "method": "textDocument/publishDiagnostics",
  "params": {
    "uri":
      "file:///Users/javi/Development/github/Foo.re",
    "diagnostics": [
      {
        "range": {
          "start": { "line": 0, "character": 0 },
          "end": { "line": 0, "character": 0 }
        },
        "severity": 1,
        "message": "invalidCharacter.orComment.orString",
        "relatedInformation": [],
        "relatedLocations": []
      }
    ]
  }
}
```

I guess it has to do with the version of the Reason parser that is included with `ocamlmerlin-reason`?

jchavarri commented:
Hm. Can't repro with `/* */` comments now. But `//` definitely make merlin unhappy.

rusty-key commented:
That's probably because I included `ocamlmerlin-reason` from `reason-cli`. I'll fix that

jchavarri commented on src/formatters/utils.ts:
```diff
@@ -0,0 +1,39 @@
+
+  if (!formatter) {
+    vscode.window.showInformationMessage(
+      `${formatterPath} is not available. Please specify "vscode-reasonml.path.${formatterName}"`,
```
> Please specify "vscode-reasonml.path.${formatterName}"

This message is not correct. The setting would be `reason.path.refmt`.

jchavarri replied:
Unrelated, but I can't get refmt to work even when using the right config.

rusty-key commented:
So, I talked to @jchavarri and we agreed that the best workflow for now is `esy` because it install all required dependencies locally and prevents global switches mess.

But I still really want to provide a seamless experience for BS users who don't want to deal with `esy`/`opam`. So, I decided to leave two options:
1. If extension detects esy configuration, it asks you to specify all required binaries such as `merlin-lsp` and `ocamlformat` as devDependencies.
2. If extension detects bsconfig without esy.json, it tries to use precompiled binaries, but still allow you to override formatters.

If we can't detect either, we just stop and ask user to provide of these configurations, because we have no idea what to do.

Next, we can provide a creation or population of esy.json and probably run esy to install dependencies. Also, I wonder, if we can hide all of this into extension internals and detect compiler version based on project configuration.

@andreypopp @jchavarri @jordwalke what do you think?

wokalski commented:
Can we perhaps make the scope of this PR more limited and merge it? Like, if there's no ocamlmerlin-lsp in path, just add an error and say that it's not in the path? It would be a great improvement itself and any follow ups would be just icing on the cake!

rusty-key commented:
@wokalski, makes sense to me. I'll prepare changes over the week

rusty-key commented:
@wokalski, it was quite a long week :)

I dumbed down the extension. Now it requires `ocamlmerlin-lsp` and `ocamlmerlin-reason` in configuration and doesn't work otherwise. Also, instead of trying to provide a working configuration, I just added instructions to README. 

I'll return to `esy`/`opam` and easier setup after this PR is merged. 

cc @andreypopp @jchavarri @jordwalke

jfrolich commented:
Is this abandoned? People were recommending `vscode-reasonml` as `reason-language-server` is pretty unreliable at the moment. This PR looks like a great improvement.

wokalski commented:
@jfrolich take a look at `ocamllabs/vscode-ocaml-platform`.
